ledger/src/bank_forks.rs

```diff
@@ -177,34 +177,55 @@ impl BankForks {
             .last()
             .map(|bank| bank.transaction_count())
             .unwrap_or(0);
-
-        root_bank.squash();
-        let new_tx_count = root_bank.transaction_count();
-
-        // Generate a snapshot if snapshots are configured and it's been an appropriate number
-        // of banks since the last snapshot
+        // Generate each snapshot at a fixed interval
+        let mut squashed_slot = None;
         if self.snapshot_config.is_some() && snapshot_package_sender.is_some() {
             let config = self.snapshot_config.as_ref().unwrap();
-            info!("setting snapshot root: {}", root);
-            if root - self.last_snapshot_slot >= config.snapshot_interval_slots as Slot {
-                let mut snapshot_time = Measure::start("total-snapshot-ms");
-                let r = self.generate_snapshot(
-                    root,
-                    &root_bank.src.roots(),
-                    snapshot_package_sender.as_ref().unwrap(),
-                );
-                if r.is_err() {
-                    warn!("Error generating snapshot for bank: {}, err: {:?}", root, r);
-                } else {
-                    self.last_snapshot_slot = root;
+            let mut banks = vec![root_bank];
+            let parents = root_bank.parents();
+            banks.extend(parents.iter());
```

Comment:
This is ordered backwards from greatest to least, which means you might squash later banks first, is this what we want?

Reply:
Yes. We want to squash the latest bank in this list that matches the interval criteria. Then we should break out. So we should only squash 1 bank in this loop, and then we will squash the bank at the passed-in slot after creating the snapshot if we didn't already.

Reply:
Doesn't this mean if you have earlier banks that also match the interval criteria, you won't generate a snapshot for them? So if one validator is making roots at > `interval`, he may not generate a snapshot for the same slots as other validators

Reply:
@carllin correct.  @sakridge i guess there is a chance they constantly skip the synchronization point

Reply:
Yea. I changed it to iterate through all the parents and start the snapshot at each interval of bank height. Now i'm just worried if it encountered say 10 squashes at once, it would just hang for a while computing the hashes of all those slots. I think that should be rare, but maybe we should do it async.

Reply:
@sakridge i think skipping is ok since we WARN on downtime, halt on inconsistency.